### PR TITLE
eventbubbling - shadowDOM

### PR DESCRIPTION
**Pull request by juancho11gm**

### CODE REVIEW

File: event-bubbling-shadowDOM/README.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+### Exercise: create shopping cart with events.
+
+* Go to the [exercise](./exercise)* folder. You will find four files (index.html, main.js, styles.css and product_class.js).
+* Create a *ShoppingCart* Class inside the modules folder. This class should have an array of *Products*. -> Done.
+* Create two methods inside the *ShoppingCart* class: *add_product(Product)* and *remove_product(product_id)*. ->Done.
+* Display the array of items of the *main.js* in the index.html like this:
+
+![shoppingcart](https://user-images.githubusercontent.com/36536646/78741526-27637d00-791f-11ea-9bd0-be2e2ab3152e.PNG)
+
+* First modidy *styles.css* and then start to modify the *main.js* file to display the products.
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/README.md:175]: typo: Modify
```suggestion
* First modify *styles.css* and then start updating the *main.js* file to display the products.
```

File: event-bubbling-shadowDOM/README.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                </div>
+            </div>
+        </div>
+    </div>
+    
+    //main.js 
+    var logElement = document.querySelector('#log'); // get log section
+    var divs = document.getElementsByTagName('div'); // get and save all the divs into an array 
+
+    for (var i = 0; i < divs.length; i++) { //loop through the array
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/README.md:88]:
```suggestion
    for (let i = 0; i < divs.length; i++) { //loop through the array
```

File: event-bubbling-shadowDOM/README.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                        <div>5</div>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+    
+    //main.js 
+    var logElement = document.querySelector('#log'); // get log section
+    var divs = document.getElementsByTagName('div'); // get and save all the divs into an array
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/README.md:86]:
```suggestion
    let logElement = document.querySelector('#log'); // get log section
    let divs = document.getElementsByTagName('div'); // get and save all the divs into an array 
```

File: event-bubbling-shadowDOM/README.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    //main.js 
+    var logElement = document.querySelector('#log'); // get log section
+    var divs = document.getElementsByTagName('div'); // get and save all the divs into an array 
+
+    for (var i = 0; i < divs.length; i++) { //loop through the array
+        divs[i].addEventListener('click', capture, true); //add capture event
+        divs[i].addEventListener('click', bubble, false); //add bubble event
+    }
+
+    var clearButton = document.querySelector('#clear'); // get clear button
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/README.md:93]:
```suggestion
    let clearButton = document.querySelector('#clear'); // get clear button
```

File: event-bubbling-shadowDOM/event-bubbling/example2/main.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+var logElement = document.querySelector('#log'); // get log section
+var divs = document.getElementsByTagName('div'); // get and save all the divs into an array 
+
+for (var i = 0; i < divs.length; i++) { //loop through the array
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/event-bubbling/example2/main.js:4]:
```suggestion
let logElement = document.querySelector('#log'); // get log section
let divs = document.getElementsByTagName('div'); // get and save all the divs into an array 

for (let i = 0; i < divs.length; i++) { //loop through the array
```

File: event-bubbling-shadowDOM/event-bubbling/example2/main.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    divs[i].addEventListener('click', capture, true); //add capture event
+    divs[i].addEventListener('click', bubble, false); //add bubble event
+}
+
+var clearButton = document.querySelector('#clear'); // get clear button
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/event-bubbling/example2/main.js:9]:
```suggestion
let clearButton = document.querySelector('#clear'); // get clear button
```

File: event-bubbling-shadowDOM/event-bubbling/example2/styles.css [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/*
+    margin: 0 auto // centrally within it's parent container.
+    display: block // It starts on a new line, and takes up the whole width
+*/
+.btn-clear {
+    margin: 0 auto;
+    display: block;
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/event-bubbling/example2/styles.css:7]:
```suggestion
    display: block;
    margin: 0 auto;
```

File: event-bubbling-shadowDOM/exercise/main.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import {Product} from './modules/product_class.js'
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/main.js:1]:
```suggestion
import { Product } from './modules/product_class.js'
```

File: event-bubbling-shadowDOM/exercise/main.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import { ShoppingCart } from './modules/shopping_cart_class.js';
+
+const div_products = document.querySelector('.products'); // products element
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/main.js:4]:
```suggestion
const productsElement = document.querySelector('.products'); // products element
```

File: event-bubbling-shadowDOM/exercise/main.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+const div_shcart = document.querySelector('.shopping-cart'); //shopping cart element
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/main.js:5]:
```suggestion
const shoppingCartElement = document.querySelector('.shopping-cart'); //shopping cart element
```

File: event-bubbling-shadowDOM/exercise/main.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+];
+
+//create a new shopping cart
+let shopping_cart = new ShoppingCart();
+
+/* Example 
+shopping_cart.add_product(items[0]);
+shopping_cart.add_product(items[1]);
+shopping_cart.add_product(items[1]);
+shopping_cart.remove_product(items[1].id);
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/main.js:22]: `items[1]` should be saved in a `const` and `id` should be destructured from that object.

File: event-bubbling-shadowDOM/exercise/modules/shopping_cart_class.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
+export { ShoppingCart };
+
+class ShoppingCart {
+    constructor() { //items = array of Products (Product.js)
+        this.items = [];
+        this.quantity = [];
+    }
+    add_product = (product) => {
+        let pos = this.getPosition(product.id); //get index of the product
+        if (pos != -1) { //if the product exists, we increase the quantity.
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/modules/shopping_cart_class.js:10]:
```suggestion
        if (pos !== -1) { //if the product exists, we increase the quantity.
```

File: event-bubbling-shadowDOM/exercise/modules/shopping_cart_class.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+            this.quantity[pos]++;
+        } else {
+            this.items.push(product); // if the product does not exists, we add it to the array
+            this.quantity.push(1);
+        }
+    }
+    remove_product = (product_id) => {
+        let pos = this.getPosition(product_id);
+        if (pos != -1) {
```
Comment (pushdev-code) [inline event-bubbling-shadowDOM/exercise/modules/shopping_cart_class.js:19]:
```suggestion
        if (pos !== -1) {
```

File: event-bubbling-shadowDOM/exercise/main.js (hunk 11 above)
Comment (juancho11gm) [reply event-bubbling-shadowDOM/exercise/main.js:22]: sure!

File: event-bubbling-shadowDOM/exercise/main.js (hunk 8 above)
Comment (juancho11gm) [reply event-bubbling-shadowDOM/exercise/main.js:1]: ok

**Pull request merged**